api/api.go

```diff
@@ -271,84 +219,91 @@ func (a *API) Series(w http.ResponseWriter, r *http.Request, _ httprouter.Params
-		return
+		return nil, nil, &thanosapi.ApiError{Typ: thanosapi.ErrorExec, Err: err}
 	}
 
-	// TODO(bwplotka): Handle warnings.
-	names, _, err := q.LabelNames()
+	names, warnings, err := q.LabelNames()
 	if err != nil {
-		a.respondPromError(w, http.StatusInternalServerError, errors.Wrap(err, "retrieve label names"))
-		return
+		return nil, nil, &thanosapi.ApiError{Typ: thanosapi.ErrorExec, Err: err}
 	}
 
-	if err = json.NewEncoder(w).Encode(LabelNamesResult{Names: names}); err != nil {
-		level.Error(a.logger).Log("msg", "error marshaling json", "err", err)
-	}
+	return names, warnings, nil
 }
 
-func (a *API) LabelValues(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+func (a *API) LabelValues(r *http.Request) (interface{}, []error, *thanosapi.ApiError) {
 	ctx := r.Context()
+	name := route.Param(ctx, "name")
 
-	name := ps.ByName("label_name")
 	if !model.LabelNameRE.MatchString(name) {
-		a.respondPromError(w, http.StatusBadRequest, errors.Errorf("invalid label name %q", name))
-	}
-
-	if err := r.ParseForm(); err != nil {
-		a.respondPromError(w, http.StatusInternalServerError, errors.Wrap(err, "parse form"))
-		return
+		return nil, nil, &thanosapi.ApiError{Typ: thanosapi.ErrorBadData, Err: errors.Errorf("invalid label name: %q", name)}
 	}
 
 	start, end, err := parseMetadataTimeRange(r, defaultMetadataTimeRange)
 	if err != nil {
-		a.respondPromError(w, http.StatusBadRequest, err)
-		return
+		return nil, nil, &thanosapi.ApiError{Typ: thanosapi.ErrorBadData, Err: err}
 	}
 
 	q, err := a.db.Querier(ctx, timestamp.FromTime(start), timestamp.FromTime(end))
 	if err != nil {
-		a.respondPromError(w, http.StatusInternalServerError, errors.Wrap(err, "new querier"))
-		return
+		return nil, nil, &thanosapi.ApiError{Typ: thanosapi.ErrorExec, Err: err}
 	}
 
-	// TODO(bwplotka): Handle warnings.
-	names, _, err := q.LabelValues(name)
+	names, warnings, err := q.LabelValues(name)
 	if err != nil {
-		a.respondPromError(w, http.StatusInternalServerError, errors.Wrap(err, "retrieve label names"))
-		return
+		return nil, nil, &thanosapi.ApiError{Typ: thanosapi.ErrorExec, Err: err}
 	}
 
-	if err = json.NewEncoder(w).Encode(LabelValuesResult{Values: names}); err != nil {
-		level.Error(a.logger).Log("msg", "error marshaling json", "err", err)
-	}
+	return names, warnings, nil
 }
 
 func (a *API) Reload(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
 	a.reloadCh <- struct{}{}
 }
+
+type ApiFunc func(w http.ResponseWriter, r *http.Request, params httprouter.Params) (interface{}, []error, *thanosapi.ApiError)
+
+// TODO: add tracer
+// Instr returns a http HandlerFunc with the instrumentation middleware.
+func Instr(
+	_ log.Logger,
+	ins extpromhttp.InstrumentationMiddleware,
+) func(name string, f thanosapi.ApiFunc) httprouter.Handle {
+	instr := func(name string, f thanosapi.ApiFunc) httprouter.Handle {
+		hf := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			thanosapi.SetCORS(w)
```

Comment:
The instr method already sets CORS, do you think it is better to remove this?

Reply:
yeah I don't think configuring cors belong here, let's remove it here